patches/packages/routing/0005-batman-adv-mesh-wide-multi-target-multicast-to-unicast-support.patch

```diff
@@ -0,0 +1,1091 @@
+++			 */
+++			kfree_skb(newskb);
+++			ret = NET_XMIT_SUCCESS;
+++		}
+++
+++		limit_tmp--;
+++	}
+++	rcu_read_unlock();
+++
+++	*limit = limit_tmp;
+++	return ret;
+++}
+++
+++/**
+++ * batadv_mcast_forw_want_all_send() - send packet to nodes in a want-all list
+++ * @bat_priv: the bat priv with all the soft interface information
+++ * @skb: the multicast packet to transmit
+++ * @vid: the vlan identifier
+++ * @limit: number of remaining, maximum transmissions
+++ *
+++ * Sends copies of a frame with multicast destination to any node with a
+++ * BATADV_MCAST_WANT_ALL_IPV4 or BATADV_MCAST_WANT_ALL_IPV6 flag set. A
+++ * transmission is performed via a batman-adv unicast packet for each such
+++ * destination node.
+++ *
+++ * Return: NET_XMIT_DROP if limit was reached, on memory allocation failure
+++ * or if the protocol family is neither IPv4 nor IPv6. NET_XMIT_SUCCESS
+++ * otherwise.
+++ */
+++static int
+++batadv_mcast_forw_want_all_send(struct batadv_priv *bat_priv,
+++				struct sk_buff *skb, unsigned short vid,
+++				unsigned int *limit)
+++{
+++	switch (ntohs(eth_hdr(skb)->h_proto)) {
+++	case ETH_P_IP:
+++		return batadv_mcast_forw_want_all_ipv4_send(bat_priv, skb, vid,
+++							    limit);
+++	case ETH_P_IPV6:
+++		return batadv_mcast_forw_want_all_ipv6_send(bat_priv, skb, vid,
+++							    limit);
+++	default:
+++		/* we shouldn't be here... */
+++		return NET_XMIT_DROP;
+++	}
+++}
+++
+++/**
+++ * batadv_mcast_forw_send() - send packet to any detected multicast recpient
+++ * @bat_priv: the bat priv with all the soft interface information
+++ * @skb: the multicast packet to transmit
+++ * @vid: the vlan identifier
+++ * @limit: number of remaining, maximum transmissions
+++ *
+++ * Sends copies of a frame with multicast destination to any node that signaled
+++ * interest in it, that is either via the translation table or the according
+++ * want-all flags. A transmission is performed via a batman-adv unicast packet
+++ * for each such destination node.
+++ *
+++ * If NET_XMIT_DROP is returned then caller needs to free the provided skb.
+++ * Otherwise it is consumed.
+++ *
+++ * Return: NET_XMIT_DROP if limit was reached, on memory allocation failure
+++ * or if the protocol family is neither IPv4 nor IPv6. NET_XMIT_SUCCESS
+++ * otherwise.
+++ */
+++int batadv_mcast_forw_send(struct batadv_priv *bat_priv, struct sk_buff *skb,
+++			   unsigned short vid)
+++{
+++	/* The previous forw mode check will try to limit to the configured
+++	 * fanout. Here, we allow a little bit of flexibility in case some
+++	 * new listeners might have joined between these function calls.
+++	 */
+++	unsigned int limit = 2 * atomic_read(&bat_priv->multicast_fanout);
+++	int ret;
+++
+++	ret = batadv_mcast_forw_tt_send(bat_priv, skb, vid, &limit);
+++	if (ret != NET_XMIT_SUCCESS)
+++		return ret;
+++
+++	ret = batadv_mcast_forw_want_all_send(bat_priv, skb, vid, &limit);
+++	if (ret != NET_XMIT_SUCCESS)
+++		return ret;
```

Comment:
In conjunction with Line 423, this seems to have the effect of not delivering the packet to nodes with the want-all flag set in case no other node has the multicast address in the transtable.

Reply:
Urgh, you are absolutely right, line 423 should return NET_XMIT_SUCCESS instead of NET_XMIT_DROP... will fix this PR later today. Many thanks for spotting and identifying the issue, @blocktrron! And thanks to you and @mweinelt for testing this PR.